### ISSUE TITLE
Background images cannot be rotated or sheered

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3874]
Images cannot be rotated, skewed, or flipped, only scaled and moved.

https://github.com/fontforge/fontforge/blob/b6a4cb45eb9234cfd3235dab1513cb624073226d/fontforge/fontviewbase.c#L486-L500

I want to fix this, but the code is quite confusing. I'm opening this issue so I remember it. Please feel free to comment if you can help, I've looked at the code for a little while and don't yet fully understand why this limitation exists.

This issue supersedes #3396.

**Comment by skef** [comment]
GImage data is stored as (horizontal, I believe) "scan lines". It's easy enough to "translate" images stored that way, and not that much trickier to scale them. Rotating or skewing them is harder. 

I don't think attempting to make the GImage code more complex for this is a good way forward. Instead you might consider adding an optional dependency on something like [libgd](https://github.com/libgd/libgd/). That way you could "export" the image data into a/the format libgd understands, run its transformation code, and reimport it. You would probably do this every time while keeping the original image data around rather than transforming and replacing it, as multiple transformations will tend to have a degraded result. 

Of course this relies on being able to [separate the rotate/skew parameters from the scale/translate parameters](https://stackoverflow.com/questions/5107134/find-the-rotation-and-skew-of-a-matrix-transformation).

**Comment by ctrlcctrlv** [comment]
Right @skef. Thanks for the libgd idea, not a bad one; I have some experience with it from many years ago when I worked on PHP code.

I guess I'd tackle this by changing the `GImage` struct...remove xoff/yoff/xscale/yscale, and add a `real transform[6]`...then figure out how to change the drawing code to handle that. Bad idea?

**Comment by skef** [comment]
I don't think that's a good idea for xoff/yoff -- you need to be able to move the image around without making it bigger or smaller and creating weird background color problems. 

You could switch to libgd (or an equivalent) for x/yscale but then you'd either need to make that dependency mandatory or make scaling optional. Probably better to take the conservative route and separate out the rotation and skewing only.

**Comment by ctrlcctrlv** [comment]
OK, I understand. So we'd want to add `real skew`, `real rotation`, and perhaps `enum flip` (`Horizontal`, `Vertical`, `Both`). For me rotation is most important, I'm often working from pages that are slightly rotated.

So what we'd need to do is...convert the transformation matrix to a single `real` describing a rotation?

**Comment by skef** [comment]
Yeah -- extract the rotation value and then use `gdImageCopyRotated`. There are separate X and Y skew functions in `gd_rotate.c` (which is *not* where `gdImageCopyRotated` is), but I don't know if those work well. 

In "traditional" transformation matrix-land flipping in dimension foo is scaling -1 in dimension foo.

**Comment by skef** [comment]
Oh , and of course it would be optimal if you can get the "fill" for a non-90-degree-multiple rotation to be transparent.

**Comment by skef** [comment]
Having thought about this more:

Let's assume that when rotating or skewing you'll want to keep the original image data around somewhere anyway. In that case it could make sense to scale with libgd when images are also being scaled or rotated. Note that this will save memory when scaling smaller but consume memory when scaling larger. It could even make sense to always scale with libgd *when it is available*, especially if libgd's scaling code does a better job -- it's worth checking. Remember, however, that making the image copy will consume memory. Basically, (assuming the offsets are handled separately) you could have a non-libgd mode that only supports scaling, and a with-libgd mode that integrates scaling, rotation, skewing, etc.

**Comment by ctrlcctrlv** [comment]
The thing that sucks about this is the SFD format. :disappointed: 

It makes the most sense to me to just add the entire transformation matrix and have FontForge figure out how to draw the image from that.

Otherwise...I have to add new tags I guess? `ImageRotation`? `ImageSkew`? Or specify new MIME types? `image/png+gd` etc which has the full `transform[6]`? Eek...

That's assuming I'm even good enough at math in the first place to convert the transformation matrix to simple degree values understandable by libgd. :nerd_face: `libgd` certainly doesn't understand our transformation matrix. I think I can handle this though, it's not that hard.

But how should I handle the file issues? Another version bump? @frank-trampe  will hate me and will surely want me to fix #3800 first.

**Comment by skef** [comment]
> That's assuming I'm even good enough at math in the first place to convert the transformation matrix to simple degree values understandable by libgd.  `libgd` certainly doesn't understand our transformation matrix. I think I can handle this though, it's not that hard.

This is why I added [this link](https://stackoverflow.com/questions/5107134/find-the-rotation-and-skew-of-a-matrix-transformation) above. 

Beyond that, what are the additional file issues? (Is the worry about saving SFDs with the broader transformation matrix values?)

**Comment by ctrlcctrlv** [comment]
It's just that we have to add rotation (and possibly skew and flip) somehow. I can probably do this just by bumping the version and tacking on a few more numbers, but it's incompatible. If I add some new tags, it's compatible, but perhaps ugly. If I'm going to add let's say rotation only since that's what's important to me, then maybe someone will come along and want to add flip as well. So it seems to make the most sense to put the entire `psMat` matrix (`transform[6]`) in the file instead

**Comment by skef** [comment]
Ah, so even though `BackgroundImageTransform()` takes a transformation matrix as parameter, that isn't stored in the file.

**Comment by ctrlcctrlv** [comment]
Right, that's what I mean; `xoff yoff xscale yscale` is what's saved. I think rotation of images is most useful for people and that's what I want to add the most. I'm open to doing flipping and skewing at the same time I work on that, but don't want to promise if it turns out difficult for some reason. I think it would be better to just store the matrix instead of add yet more values...

**Comment by ctrlcctrlv** [comment]
Right now what I do when I'm recreating a font from a manuscript is I have to first rotate the page in (e.g.) the GIMP. I always think, "why can't I just do this in FontForge?" It's highly unintuitive that the rotate seems to do _something_, it scales and moves the image in a weird way. So that's why I want to fix this, to make it more intuitive for users (I'm not the only one who has noticed this limitation over the years, see e.g. #3396, and I can't find it ATM but someone has also complained on the ML), and to make it easier for myself. :-)

**Comment by skef** [comment]
Well `xoff` and `yoff` are more or less values `M[4]` and `M[5]` respectively. And `xscale` is basically `M[0]` with `yscale` being basically `M[3]`. So perhaps the line could have a variant with two additional values corresponding to `M[1]` and `M[2]` tacked on to the end.

**Comment by ctrlcctrlv** [comment]
Maybe what I can do is only support this with the new `Image2` tag, which was designed to be the last image tag ever needed. To say that I have two more values, I'll use the `+` MIME helpfully provides, so `image/png+gd`. I'll make FontForge always write the entire matrix, and emit a warning if FontForge doesn't have `gd` but it thinks the matrix requires it for correct display.

**Comment by skef** [comment]
This note may not belong here, but when I was looking at the sample SFD you sent me recently I realized that images aren't hidden when you hide the layer (with the eye symbol). That seems wrong.

**Comment by ctrlcctrlv** [comment]
@skef Are you on the background layer when you click the eye? FontForge never hides a layer you're on.

**Comment by skef** [comment]
Oh, that does work. I thought I had tried that but I guess not.

**Comment by frank-trampe** [comment]
If it's what it takes to do this right, I'd be okay with an "Image3" tag.

**Comment by ctrlcctrlv** [comment]
@frank-trampe Thinking about this more, I'd like to propose an `Image2Metadata` tag. Just like the `Image2` tag, the first parameter is a magic key, except this time not a MIME type, but rather just ASCII [a-zA-Z0-9] describing the type of metadata.

So for this request it'd be:

```sfd
Image2Metadata: transformmatrix6 0.0 0.0 0.0 0.0 0.0 0.0
```

This way we wouldn't disturb the current xoff/yoff/xscale/yscale, and would still write `Image` and `Image2` as before. If a new FontForge compiled with `gd` saw the `Image2Metadata` tag directly after an `Image2`, it would act appropriately, but old FontForge would ignore it.

`Image3` seems...well, I'll do it if that's what consensus coalesces around. But we literally just added `Image2`!

**Comment by frank-trampe** [comment]
There are indeed some compatibility benefits to that, and I guess we still aren't sure whether we're done with image extensions, so probably a good approach.

**Comment by ctrlcctrlv** [comment]
I determined Cairo can do this, no reason to involve `gd`, while working on fixing #3919, which I failed at doing, but working on it got me reading Cairo docs!
